core/arch/arm/mm/tee_mmu.c

```diff
@@ -687,6 +687,7 @@ static TEE_Result tee_mmu_user_va2pa_attr(const struct user_ta_ctx *utc,
 				if (res != TEE_SUCCESS)
 					return res;
 
+				p = ROUNDDOWN(p, SMALL_PAGE_SIZE);
```

Comment:
If `(paddr_t)ua - utc->mmu->regions[n].va` was added to `utc->mmu->regions[n].offset` above we should get the correct physical address in `p` directly.

How this can work with physically noncontiguous mobjs I don't understand.

Reply:
The mobj_get_pa should work for noncontiguous buffers.
I will try add this test in the xtest i'm using.

Reply:
yep, my fault; i must be right. we must provide the real target offset to mobj_get_pa .

Reply:
Even with `(paddr_t)ua - utc->mmu->regions[n].va` > `SMALL_PAGE_SIZE` and the mobj is truly physically noncontiguous?

Reply:
(ua - region->va) is the offset from mobj base mapping to target address. Should use this and get pa with small page granule.
Then from small page align phys addr, we add the input `ua` (user vaddr) offset in small page alignment.